exchanges/bitfinex/bitfinex_websocket.go

```diff
@@ -869,139 +873,99 @@ func (b *Bitfinex) handleWSTickerUpdate(c *subscription.Subscription, d []interf
-				}
-				wsTrade.Rate = rate
-				period, ok := elem[4].(float64)
-				if !ok {
-					return errors.New("unable to type assert trade period")
-				}
-				wsTrade.Period = int64(period)
-			} else {
-				price, ok := elem[3].(float64)
-				if !ok {
-					return errors.New("unable to type assert trade price")
-				}
-				wsTrade.Rate = price
-			}
-			tradeHolder = append(tradeHolder, wsTrade)
-		}
-	case 3:
-		if eventType != wsFundingTradeUpdate && eventType != wsTradeExecutionUpdate {
-			return fmt.Errorf("unhandled WS trade update event: %s", eventType)
-		}
-		data, ok := d[2].([]interface{})
-		if !ok {
-			return errors.New("trade data type assertion error")
+	var wsTrades []*wsTrade
+	switch valueType {
+	case jsonparser.String:
+		if t, err := b.handleWSPublicTradeUpdate(respRaw); err != nil {
+			return err
+		} else {
+			wsTrades = []*wsTrade{t}
 		}
-
-		tradeID, ok := data[0].(float64)
-		if !ok {
-			return errors.New("unable to type assert trade ID")
+	case jsonparser.Array:
+		if wsTrades, err = b.handleWSPublicTradesSnapshot(respRaw); err != nil {
+			return err
 		}
-		timestamp, ok := data[1].(float64)
-		if !ok {
-			return errors.New("unable to type assert trade timestamp")
+	default:
+		return fmt.Errorf("%w `tradesUpdate[1]`: %w `%s`", errParsingWSField, jsonparser.UnknownValueTypeError, valueType)
+	}
+	trades := make([]trade.Data, len(wsTrades))
+	for i, w := range wsTrades {
+		t := trade.Data{
+			Exchange:     b.Name,
+			AssetType:    s.Asset,
+			CurrencyPair: s.Pairs[0],
+			TID:          strconv.FormatInt(w.ID, 10),
+			Timestamp:    w.Timestamp.Time().UTC(),
+			Side:         order.Buy,
+			Amount:       w.Amount,
+			Price:        w.Price,
 		}
-		amount, ok := data[2].(float64)
-		if !ok {
-			return errors.New("unable to type assert trade amount")
+		if w.Period != 0 {
+			t.AssetType = asset.MarginFunding
 		}
-		wsTrade := WebsocketTrade{
-			ID:        int64(tradeID),
-			Timestamp: int64(timestamp),
-			Amount:    amount,
+		if t.Amount < 0 {
+			t.Side = order.Sell
+			t.Amount = math.Abs(t.Amount)
 		}
-		if len(data) == 5 {
-			rate, ok := data[3].(float64)
-			if !ok {
-				return errors.New("unable to type assert trade rate")
-			}
-			period, ok := data[4].(float64)
-			if !ok {
-				return errors.New("unable to type assert trade period")
-			}
-			wsTrade.Rate = rate
-			wsTrade.Period = int64(period)
+		trades[i] = t
+	}
+	if feedEnabled {
+		b.Websocket.DataHandler <- trades
+	}
+	if b.IsSaveTradeDataEnabled() {
+		err = trade.AddTradesToBuffer(b.GetName(), trades...)
+	}
+	return err
+}
+
+func (b *Bitfinex) handleWSPublicTradesSnapshot(respRaw []byte) (trades []*wsTrade, errs error) {
+	handleTrade := func(v []byte, valueType jsonparser.ValueType, _ int, _ error) {
+		if valueType != jsonparser.Array {
+			errs = common.AppendError(errs, fmt.Errorf("%w `tradesSnapshot[1][*]`: %w `%s`", errParsingWSField, jsonparser.UnknownValueTypeError, valueType))
```

Comment:
This is a little bit extra given that there is only one caller to this and its behind a gate checking if its an array. I am normally one for all checks should be within the function, but its that you just checked to get to this function

Reply:
How about this simplification?
```go
func (b *Bitfinex) handleWSPublicTradesSnapshot(respRaw []byte) ([]*wsTrade, error) {
	var trades []*wsTrade
	err := json.Unmarshal(respRaw, &trades)
	return trades, err
}
```
Where `handleWSAllTrades` changes with:
```go
k, valueType, _, err := jsonparser.Get(respRaw, "[1]")
...
if wsTrades, err = b.handleWSPublicTradesSnapshot(k); err != nil {
```
A lot less tangling of additional jsonparser functions.

Reply:
> This is a little bit extra given that there is only one caller to this and its behind a gate checking if its an array. 

I think you might have missed that this is checking that every element inside the array is also an array.
That's the answer to why we're doing the second check.

And yeah. I think I didn't re-sniff simplifying this after applying shazbert's suggestion of the UnmarshalJSON on wsTrade. I think we can probably do exactly as you suggest. Will fix the test and confirm.

Reply:
Yep. Exactly as you suggested. Extended to the handleWSPublicTradeUpdate as well.

Fixed https://github.com/gbjk/gocryptotrader/commit/82eb2c02d